modular_skyrat/modules/ashwalkers/code/effects/ash_rituals.dm

```diff
@@ -104,17 +104,24 @@
 	for(var/mob/select_mob in GLOB.player_list)
 		if(select_mob.z != success_rune.z)
 			continue
+
 		to_chat(select_mob, span_userdanger("The planet stirs... another monster has arrived!"))
 		playsound(get_turf(select_mob), 'sound/magic/demon_attack1.ogg', 50, TRUE)
 		flash_color(select_mob, flash_color = "#FF0000", flash_time = 3 SECONDS)
+
 	var/megafauna_choice = pick(
 		/mob/living/simple_animal/hostile/megafauna/blood_drunk_miner,
 		/mob/living/simple_animal/hostile/megafauna/dragon,
 		/mob/living/simple_animal/hostile/megafauna/hierophant,
 	)
+
 	var/turf/spawn_turf = locate(rand(1,255), rand(1,255), success_rune.z)
-	while(!istype(spawn_turf, /turf/open/misc/asteroid/basalt))
+
+	var/anti_endless = 0
+	while(!istype(spawn_turf, /turf/open/misc/asteroid) && anti_endless < 100)
 		spawn_turf = locate(rand(1,255), rand(1,255), success_rune.z)
+		anti_endless++
+
 	new /obj/effect/particle_effect/sparks(spawn_turf)
 	sleep(3 SECONDS)
```

Comment:
Would've been nice to have used `addtimer()` here instead.

Reply:
other while loops have done this to prevent an infinite loop-- I'm trying to understand what you want me to do with mentioning the addtimer.

Reply:
The `sleep(3 SECONDS)` should instead be an `addtimer()` with a time of `3 SECONDS` and a callback to a proc that runs the `new megafauna_choice(spawn_turf)` (might need to give it the `spawn_turf` as an argument, too), as it's better practice to use those rather than sleeps, so that they can respect the master controller and can show up in the profiler, instead of us having no way to control them whatsoever.

Reply:
OH, see, github things. I thought you meant to add a timer for the anti_endless thing and I was massively confused.